crates/node/src/shell/finalize_block.rs

```diff
@@ -5796,4 +5796,165 @@ mod test_finalize_block {
+                    Fee {
+                        amount_per_gas_unit: DenominatedAmount::native(
+                            1.into(),
+                        ),
+                        token: shell.state.in_mem().native_token.clone(),
+                    },
+                    sk.ref_to(),
+                    WRAPPER_GAS_LIMIT.into(),
+                ))));
+            batch.header.chain_id = shell.chain_id.clone();
+            batch.header.atomic = false;
+
+            // append first inner tx to batch
+            batch.set_code(Code::new(TestWasms::TxNoOp.read_bytes(), None));
+            batch.set_data(Data::new("bing".as_bytes().to_owned()));
+
+            // append second inner tx to batch
+            batch.push_default_inner_tx();
+
+            batch.set_code(Code::new(TestWasms::TxNoOp.read_bytes(), None));
+            batch.set_data(Data::new("bong".as_bytes().to_owned()));
+
+            // sign the batch of txs
+            batch.sign_raw(
+                vec![sk.clone()],
+                vec![sk.ref_to()].into_iter().collect(),
+                None,
+            );
+            batch.sign_wrapper(sk);
+
+            batch
+        };
+
+        let processed_txs = vec![ProcessedTx {
+            tx: batch_tx.to_bytes().into(),
+            result: TxResult {
+                code: ResultCode::Ok.into(),
+                info: "".into(),
+            },
+        }];
+
+        let mut events = shell
+            .finalize_block(FinalizeBlock {
+                txs: processed_txs,
+                ..Default::default()
+            })
+            .expect("Test failed");
+
+        // one top level event
+        assert_eq!(events.len(), 1);
+        let event = events.remove(0);
+
+        // multiple tx results (2)
+        let tx_results = event.read_attribute::<Batch<'_>>().unwrap();
+        assert_eq!(tx_results.batch_results.0.len(), 2);
+
+        // all txs should have succeeded
+        assert!(
+            tx_results
+                .batch_results
+                .0
+                .values()
+                .all(|result| result.is_ok())
+        );
+    }
+
+    #[test]
+    fn test_multiple_events_from_batch_tx_one_valid_other_invalid() {
+        let (mut shell, _, _, _) = setup();
+
+        let sk = wallet::defaults::bertha_keypair();
+
+        let batch_tx = {
+            let mut batch =
+                Tx::from_type(TxType::Wrapper(Box::new(WrapperTx::new(
+                    Fee {
+                        amount_per_gas_unit: DenominatedAmount::native(
+                            1.into(),
+                        ),
+                        token: shell.state.in_mem().native_token.clone(),
+                    },
+                    sk.ref_to(),
+                    WRAPPER_GAS_LIMIT.into(),
+                ))));
+            batch.header.chain_id = shell.chain_id.clone();
+            batch.header.atomic = false;
+
+            // append first inner tx to batch (this one is valid)
```

Comment:
Nit, we have a `mk_tx_batch` helper for unit tests in `finalize_block` that you could use to build the batch